text/0000-undo-universal-impl-trait.md

```diff
@@ -0,0 +1,238 @@
+return location for a while now (mostly on nightly) and the initial RFC was well accepted among the
+community and was a success. Those developers obviously mapped `impl Trait` to *callee-picked*
+types, because at that time it was its sole semantics. When a function with an argument annotated
+with `impl Trait` appears to such a developer, they get confused because that would mean *an argument
+which type is chosen by the callee*, which doesn’t quite make sense.
+
+### Argument from ergonomics
+
+> Argument [here](https://github.com/rust-lang/rfcs/blob/master/text/1951-expand-impl-trait.md#argument-from-ergonomics)
+
+In this argument, [RFC 1951] states that having `impl Trait` in argument position eases how our
+mental model about a function works. We don’t have to read the type variables and rembember their
+trait bounds anymore to understand the contract of a function.
+
+This argument seems sound but again is wrong. Consider:
+
+```
+fn foo<T>(a: T, b: T, c: impl Debug) where T: Add<Output = T> + Copy
+```
+
+This function has three type variables to substitute to be completely monomorphized. However, it’s
+very hard to see it at first because one of the variables is hidden. Worse, the ergonomics argument
+doesn’t stand here because you can see you **have** to read the type variables in order to get what
+you can do with the variables.
+
+Also, consider this function that takes an argument and add it to itself:
+
+```
+fn add_self<T>(x: T) -> T where T: Add<Output = T> + Copy
+```
+
+Now consider:
+
+```
+fn add_self(x: impl Add<Output = impl Copy> + Copy)
+```
+
+You can see the duplication here and the overwhelming confusion that both the `impl Trait` will
+resolve to the same type, even though they don’t have the same contract, which is impossible to
+guess from the interface while it could. This is legal because we only talk about two contracts here
+and the function will pick a type at the union (it must be `Add + Copy` and `Copy`) but you also
+have that weird `Output` contract as well.
+
+Even weirder:
+
+```
+fn add_self(x: impl Add<Output = impl Sized> + Copy) -> impl Sized {
+  x + x
+}
+```
+
+### Argument from familiarity
+
+> Argument [here](https://github.com/rust-lang/rfcs/blob/master/text/1951-expand-impl-trait.md#argument-from-familiarity)
+
+This argument is quite the same as the one for ergonomics and yet has this sentence:
+
+> In Rust, this is just a syntactic easement into a unitary polymorphism system which is
+> fundamentally one idea: parametric polymorphism with trait constraints.
+
+The example given in that statement is itself a proof that *unitary polymorphism* isn’t that common
+since the `U` type variable *needs* to be explicitely written. The familiarity with Java or C# might
+hold but shouldn’t drive the design decision and even more important, the consistency of the Rust
+syntax.
+
+### Summary of arguments
+
+This document has shown that `impl Trait` in argument position is confusing for newcomers and
```

Comment:
This document hasn't really shown *anything* from newcomers. It's pretty hard for us to gauge how this affects newcomers at this point, since it does not appear in the book or anywhere in the documentation.

Instead this document seems to be claiming every other argument is subjective, but this one is somehow objective.

Perhaps the source of confusion is the fact that it's a brand new, undocumented feature that people are still just starting to use.

Reply:
I’ve linked a few links in the top of the document. Also, reactions and the fact it’s been so opinionated is somehow a hint that there’s a massive confusion, right?

Reply:
I mean, personally I was confused as to why we had the same syntax for the same thing, but that's just me I guess.

Reply:
It’s not just you. All of this is very confusing and opinionated.

Reply:
> Also, reactions and the fact it’s been so opinionated is somehow a hint that there’s a massive confusion, right?

People having strong opinions does not mean they're confused.